ledger/complete/compactor.go

```diff
@@ -0,0 +1,393 @@
+	"github.com/onflow/flow-go/module/observable"
+)
+
+type WALTrieUpdate struct {
+	Update   *ledger.TrieUpdate
+	ResultCh chan<- error
+	DoneCh   <-chan struct{}
+}
+
+type checkpointResult struct {
+	num int
+	err error
+}
+
+type Compactor struct {
+	checkpointer       *realWAL.Checkpointer
+	wal                *realWAL.DiskWAL
+	ledger             *Ledger
+	logger             zerolog.Logger
+	stopCh             chan chan struct{}
+	trieUpdateCh       <-chan *WALTrieUpdate
+	trieUpdateDoneCh   chan<- struct{}
+	lm                 *lifecycle.LifecycleManager
+	observers          map[observable.Observer]struct{}
+	checkpointDistance uint
+	checkpointsToKeep  uint
+}
+
+func NewCompactor(l *Ledger, w *realWAL.DiskWAL, checkpointDistance uint, checkpointsToKeep uint, logger zerolog.Logger) (*Compactor, error) {
+	if checkpointDistance < 1 {
+		checkpointDistance = 1
+	}
+
+	checkpointer, err := w.NewCheckpointer()
+	if err != nil {
+		return nil, err
+	}
+
+	trieUpdateCh := l.TrieUpdateChan()
+	if trieUpdateCh == nil {
+		return nil, errors.New("failed to get valid trie update channel from ledger")
+	}
+
+	trieUpdateDoneCh := l.TrieUpdateDoneChan()
+	if trieUpdateDoneCh == nil {
+		return nil, errors.New("failed to get valid trie update done channel from ledger")
+	}
+
+	return &Compactor{
+		checkpointer:       checkpointer,
+		wal:                w,
+		ledger:             l,
+		logger:             logger,
+		stopCh:             make(chan chan struct{}),
+		trieUpdateCh:       trieUpdateCh,
+		trieUpdateDoneCh:   l.trieUpdateDoneCh,
+		observers:          make(map[observable.Observer]struct{}),
+		lm:                 lifecycle.NewLifecycleManager(),
+		checkpointDistance: checkpointDistance,
+		checkpointsToKeep:  checkpointsToKeep,
+	}, nil
+}
+
+func (c *Compactor) Subscribe(observer observable.Observer) {
+	var void struct{}
+	c.observers[observer] = void
+}
+
+func (c *Compactor) Unsubscribe(observer observable.Observer) {
+	delete(c.observers, observer)
+}
+
+// Ready periodically fires Run function, every `interval`
+func (c *Compactor) Ready() <-chan struct{} {
+	c.lm.OnStart(func() {
+		go c.run()
+	})
+	return c.lm.Started()
+}
+
+func (c *Compactor) Done() <-chan struct{} {
+	c.lm.OnStop(func() {
+		// Close trieUpdateDoneCh to signal trie updates are finished
+		defer close(c.trieUpdateDoneCh)
+
+		// Notify observers
+		for observer := range c.observers {
+			observer.OnComplete()
+		}
+
+		// Signal Compactor goroutine to stop
+		doneCh := make(chan struct{})
+		c.stopCh <- doneCh
+
+		// Wait for Compactor goroutine to stop
+		<-doneCh
+	})
+	return c.lm.Stopped()
+}
+
+func (c *Compactor) run() {
+
+	// checkpointSem is used to limit checkpointing to one.
+	// If previous checkpointing isn't finished when enough segments
+	// are finalized for next checkpointing, retry checkpointing
+	// again when next segment is finalized.
+	// This avoids having more tries in memory than needed.
+	checkpointSem := semaphore.NewWeighted(1)
```

Comment:
I don’t understand how this works. Every call will create a new semaphore object. Wouldn’t “TryAcquire” always return true? Don’t we need to share the semaphore object between different calls?

Reply:
> I don’t understand how this works.  Every call will create a new semaphore object. 

No, that is incorrect.  `checkpointSem` is initialized outside of goroutine's `for` loop which waits for and processes messages from channels.  So it is created once and used inside `for` loop to limit one checkpointing at a time. 

> Wouldn’t “TryAcquire” always return true? Don’t we need to share the semaphore object between different calls?

No, `TryAquire` doesn't always return true.  All checkpointings need to acquire the same semaphore.  So this limits to one checkpointing at a time, and all other calls to `TryAcquire` returns false immediately (non-blocking).